Shopify/cli
# Generate app schema dynamically

### WHY are these changes introduced?


Follow-up: https://github.com/Shopify/cli/pull/3188

NOTES:

- Rebase follow-up branch before merging it

The app configuration is loaded from the current fixed `AppSchema`. Some parts of this schema are loaded as `configuration` extensions. In case, any of the `configuration` extension specification is configured behind a beta flag, that part of the configuration is not sent (the app should be loaded using the `remote specs` check the previous [PR](https://github.com/Shopify/cli/pull/3180)`
However, the user is not receiving any error when the app is loaded and the config content includes a section associated with a gated `configuration` spec


### WHAT is this pull request doing?
- App loader generates the `app schema` dynamically from a list of `configuration` extension specifications
- The App loader can receive the list of extension specifications
- In case the list of specs is empty, the loader uses the local ones (backwards compatibility with some parts of the code that needs to load the configuration but they are not passing the new `specification` parameter`)
- Configuration extension schemas have been moved to the spec definition. All of them have been removed fro the `AppSchema`
- Each time the CLI needs to load/write the configuration, the `versioned app schema` should be generated with:
  - Non versioned fields
  - The union of every configuration extension schema


### How to test your changes?
- Create a new `cli` app and run the `app config link` command to link it to a new remote app. 

- With the org beta flag `App Access Module` disabled
  - Add the section `access` with whatever value 
  - Run the `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP` command
  - An error should be displayed saying that the `access` section is not allowed
![image](https://github.com/Shopify/cli/assets/105213827/9c599285-27a3-4bc4-acd1-f10c2644dd39)

- With the org beta flag `App Access Module` enabled
  - Add the section `access` with whatever value 
  - Run the `SHOPIFY_CLI_VERSIONED_APP_CONFIG=1 npm run shopify app deploy --path /PATH/TO/YOUR/APP` command
  - Modify the value inside the `access` section
  - Run the `p shopify app config link --path /PATH/TO/YOUR/APP` command. The `access` section content will be restored to the content it had before running the `deploy` command as it was uploaded with it


### Post-release steps



### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
